File: JavaScript/001/Elixir-quiz/day3-part1.js
import fs from 'fs'

const data = fs.readFileSync('input-part3.txt').toString();

const dataList = data.trim().split('\n');

let initValue = [];
dataList[0].split('').forEach((value, index) => {
    console.log(index)
    initValue.push({zero : 0, one : 0});
})

console.log(initValue)
const totalCount = dataList.reduce((acc, cur, i) => {
    const charList = cur.split('')
    acc = charList.reduce((accChar, cur, i) => {
        if(cur == '0') {
            accChar[i].zero++;
        }
        else if (cur == '1') {
            accChar[i].one++;
        }
        console.log(accChar[i])
        return accChar;
    }
    , acc)

    return acc;
}, initValue)

const up = totalCount.reduce((acc, cur, i) => {
    acc += cur.zero > cur.one? '1' : '0'
    return acc
}, '')

const down = totalCount.reduce((acc, cur, i) => {
    acc += cur.zero > cur.one? '0' : '1'
    return acc;
},'')

function changeBase(number, fromBase, toBase) {
    if (fromBase == 10)
        return (parseInt(number)).toString(toBase)
    else if (toBase == 10)
        return parseInt(number, fromBase);
    else {
        var numberInDecimal = parseInt(number, fromBase);
        return parseInt(numberInDecimal).toString(toBase);
    }
}

const number1 = changeBase(up, 2, 10);
const number2 = changeBase(down, 2, 10);
console.log(number1 * number2)

File: Elixir/day3-part1.exs
{:ok, file} = File.read("input-part3.txt")

lines = String.split(file, "\n", trim: true)

counting =
  1..String.length(hd(lines))
  |> Enum.reduce(%{}, fn item, acc ->
    Map.put(acc, item - 1, %{zero: 0, one: 0})
  end)

foo =
  String.split(file, "\n", trim: true)
  |> Enum.reduce(counting, fn item, acc ->
    item
    |> String.graphemes()
    |> Enum.with_index()
    |> Enum.reduce(acc, fn
      {"1", index}, chAcc ->
        Map.update!(chAcc, index, fn %{one: one, zero: zero} -> %{one: one + 1, zero: zero} end)

      {"0", index}, chAcc ->
        Map.update!(chAcc, index, fn %{one: one, zero: zero} -> %{one: one, zero: zero + 1} end)
    end)
  end)


up =
  Enum.map(foo, fn {_, %{one: one, zero: zero}} ->
    cond do
      zero > one -> "0"
      one > zero -> "1"
    end
  end)
  |>Enum.join()
  |>String.to_integer(2)
  |>IO.inspect()

  down =
    Enum.map(foo, fn {_, %{one: one, zero: zero}} ->
      cond do
        zero > one -> "1"
        one > zero -> "0"
      end
    end)
    |>Enum.join()
    |> String.to_integer(2)
    |>IO.inspect()

  # String.to_integer(up, 2)
  # |>IO.inspect()
  #  * String.to_integer(down, 2)
  # |> IO.inspect()
